py3tftp/protocols.py

```diff
@@ -291,6 +298,60 @@ def datagram_received(self, data, addr):
             logging.debug('Ack: {0}; is_ack: {1}; counter: {2}'.format(
                 data, packet.is_ack(), self.counter))
 
+    def is_packet_inside_window(self, packet, windowsize):
+        return ((packet.block_no > (self.counter - windowsize)) and (
+            packet.block_no <= self.counter))
+
+    def datagram_received_windowsize(self, data, addr, windowsize):
+        """
+        Checks correctness of incoming datagrams, reset timers,
+        increments message counter, send next chunk of requested file
+        to client, and according to the agreed windowsize.
+        """
+        packet = self.packet_factory.from_bytes(data)
+        if (self.is_correct_tid(addr) and packet.is_ack() and
+                    self.is_packet_inside_window(packet, windowsize)):
+            self.conn_timeout_reset()
+            [retransmit.cancel() for retransmit in self.retransmits]
```

Comment:
I noticed that in some cases, the connection will timeout and call `self.conn_timeout()`, which will clean up the connection. However, there will still be some retransmits in `self.retransmits` which will try to use the now closed connection, fail, and raise an error:

```
2018-09-02 15:30:04,553 [ERROR] Connection to ('127.0.0.1', 39841) timed out, "b'LICENSE'" not transfered
2018-09-02 15:30:04,553 [INFO] Connection to 127.0.0.1:39841 terminated
2018-09-02 15:30:04,556 [ERROR] Exception in callback BaseTFTPProtocol.reply_to_client(b'\x00\x03\x0...yright notice')
handle: <TimerHandle when=5553.010373637 BaseTFTPProtocol.reply_to_client(b'\x00\x03\x0...yright notice')>
Traceback (most recent call last):
  File "/home/matt/.pyenv/versions/3.5.4/lib/python3.5/asyncio/selector_events.py", line 1102, in sendto
    self._sock.sendto(data, addr)
AttributeError: 'NoneType' object has no attribute 'sendto'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/matt/.pyenv/versions/3.5.4/lib/python3.5/asyncio/events.py", line 127, in _run
    self._callback(*self._args)
  File "/home/matt/projects/py3tftp/py3tftp/protocols.py", line 149, in reply_to_client
    self.transport.sendto(packet, self.remote_addr)
  File "/home/matt/.pyenv/versions/3.5.4/lib/python3.5/asyncio/selector_events.py", line 1111, in sendto
    'Fatal write error on datagram transport')
  File "/home/matt/.pyenv/versions/3.5.4/lib/python3.5/asyncio/selector_events.py", line 639, in _fatal_error
    self._loop.call_exception_handler({
AttributeError: 'NoneType' object has no attribute 'call_exception_handler'
2018-09-02 15:30:04,559 [ERROR] Exception in callback BaseTFTPProtocol.reply_to_client(b'\x00\x03\x0...WITH THE SOFT')
handle: <TimerHandle when=5553.010453728 BaseTFTPProtocol.reply_to_client(b'\x00\x03\x0...WITH THE SOFT')>
Traceback (most recent call last):
  File "/home/matt/.pyenv/versions/3.5.4/lib/python3.5/asyncio/selector_events.py", line 1102, in sendto
    self._sock.sendto(data, addr)
AttributeError: 'NoneType' object has no attribute 'sendto'
```

I think that inserting the same line - `[retransmit.cancel() for retransmit in self.retransmits]`  - around line 189 (conn_timeout method, between calls to self.retransmit_reset and self.transport.close), prevents this. Eg.

```python
self.retransmit_reset()
[retransmit.cancel() for retransmit in self.retransmits]
self.transport.close()
```
Do you think you could add this bit there?

Reply:
I added a test to RFC7440 that reproduces the error and change the code to fix the error.